### PR TITLE
Create a caching solution for quick subsequent paints

### PR DESCRIPTION
**Pull request by wilsonpage**

### CODE REVIEW

**Comment by wilsonpage** [comment]
@etiennesegonzac r?

**Comment by wilsonpage** [comment]
This is a V1. V2 will come with height caching so that the cached render is scrollable instantly.

File: fast-list.js [hunk 1]
```diff
@@ -287,6 +294,16 @@ FastList.prototype = {
...
 
+  replaceChild: function(replacement, child) {
+    debug('replace child', replacement, child);
+    var itemsInDOM = this.els.itemsInDOM;
+    var items = this.els.items;
+
+    items[items.indexOf(child)] = replacement;
+    itemsInDOM[itemsInDOM.indexOf(child)] = replacement;
+    this.els.itemContainer.replaceChild(replacement, child);
+  },
```
Comment (wilsonpage) [inline fast-list.js:308]: This is a new public method that allows users to replace a list-item node.

File: fast-list.js [hunk 2]
```diff
@@ -313,14 +332,14 @@ FastList.prototype = {
     var headerHeight = source.getSectionHeaderHeight();
     var sections = this.source.getSections();
 
-    sections.forEach(function(section, i) {
-      var height = source.getFullSectionHeight(section);
+    for (var j = 0; j < sections.length; j++) {
```
Comment (wilsonpage) [inline fast-list.js:338]: Micro optimization

File: fast-list.js [hunk 3]
```diff
@@ -41,7 +41,9 @@ function FastList(source) {
     itemContainer: source.itemContainer || source.list,
     container: source.container,
     list: source.list,
-    sections: []
+    sections: [],
+    items: source.items || [],
+    itemsInDOM: [].concat(source.items || [])
```
Comment (wilsonpage) [inline fast-list.js:46]: For consistency we keep all element references under the `this.els` namespace.

File: gaia-fast-list.js [hunk 4]
```diff
@@ -82,11 +96,25 @@ var GaiaFastListProto = {
...
+      }
+
+      .fast-list.empty {
+        background: repeating-linear-gradient(
+          0deg,
+          var(--background) 0px,
+          var(--background) 59px,
+          var(--border-color) 60px,
+          var(--border-color) 60px);
+        background-position: 0 1px;
```
Comment (wilsonpage) [inline gaia-fast-list.js:117]: Added some fake list item borders that can be painted when the cache empty (first render).

File: gaia-fast-list.js [hunk 5]
```diff
@@ -95,6 +123,20 @@ var GaiaFastListProto = {
...
+        position: -webkit-sticky;
+        top: 0px;
+        height: 32px;
+        z-index: 100;
+
+        display: block;
+
+        background: var(--background-plus);
+        color: var(--title-color);
+      }
```
Comment (wilsonpage) [inline gaia-fast-list.js:138]: Fake section header to be painted when cache is empty (first render).

File: gaia-fast-list.js [hunk 6]
```diff
@@ -138,6 +180,11 @@ var GaiaFastListProto = {
         background: var(--background);
         -moz-user-select: none;
         text-decoration: none;
+        will-change: initial !important;
+      }
+
+      :host(.layerize) ::content .gfl-item {
+        will-change: transform !important;
```
Comment (wilsonpage) [inline gaia-fast-list.js:187]: Only apply costly `will-change` property when the component had the `.layerize` class.

File: gaia-fast-list.js [hunk 7]
```diff
@@ -138,6 +180,11 @@ var GaiaFastListProto = {
         background: var(--background);
         -moz-user-select: none;
         text-decoration: none;
+        will-change: initial !important;
```
Comment (wilsonpage) [inline gaia-fast-list.js:183]: Turned off by default.

File: gaia-fast-list.js [hunk 8]
```diff
@@ -204,29 +252,39 @@ Internal.prototype = {
...
   },
 
   sectionize: function(items) {
     var hash = {};
+
     if (!this.getSectionName) { return hash; }
 
-    items.forEach(item => {
-      var section = this.getSectionName(item);
+    for (var i = 0, l = items.length; i < l; i++) {
```
Comment (wilsonpage) [inline gaia-fast-list.js:273]: Small optimization.

File: gaia-fast-list.js [hunk 9]
```diff
@@ -241,12 +299,19 @@ Internal.prototype = {
...
-    if (templateHeader) this.templateHeader = templateHeader.innerHTML;
-    if (templateItem) this.templateItem = templateItem.innerHTML;
+    if (templateHeader) {
+      this.templateHeader = templateHeader.innerHTML;
+      templateHeader.remove();
+    }
+
+    if (templateItem) {
+      this.templateItem = templateItem.innerHTML;
+      templateItem.remove();
```
Comment (wilsonpage) [inline gaia-fast-list.js:309]: `<template>`s aren't needed after this point.

File: gaia-fast-list.js [hunk 10]
```diff
@@ -264,12 +329,20 @@ Internal.prototype = {
...
   populateItem: function(el, i) {
-    poplar.populate(el, this.getRecordAt(i));
+    // debug('populate item', el, i);
+    var record = this.getRecordAt(i);
+    if (!poplar.populate(el, record)) {
+      debug('not a poplar element');
+      var replacement = this.fastList.createItem();
+      this.fastList.replaceChild(replacement, el);
+      this.populateItem(replacement, i);
+    }
```
Comment (wilsonpage) [inline gaia-fast-list.js:357]: Poplar now tells us if populate was successful or not by returning a boolean. If unsuccessful we assume that the element passed was not a poplar element and must be replaced with a new poplar element that we can template.

File: gaia-fast-list.js [hunk 11]
```diff
@@ -338,7 +416,7 @@ Internal.prototype = {
   },
 
   getIndexAtPosition: function(pos) {
-    debug('get index at position', pos);
+    // debug('get index at position', pos);
```
Comment (wilsonpage) [inline gaia-fast-list.js:431]: Commented out to reduce noise in debug mode.

**Comment by arcturus** [comment]
Quick question, will this PR allow us to save elements on the dom list to be populated before the datasource is ready?

**Comment by wilsonpage** [comment]
@arcturus 'save elements on the dom list'?

This PR caches HTML on initial list render so that we can inject HTML instantly on next initialization, updating the cached render once the model finally arrives.

**Comment by wilsonpage** [comment]
Right now if the user provides list-item elements in their configuration they are not populated on initialization, only positioned. This meets the requirements of the `<gaia-fast-list>` use-case, but may not be appropriate for other cases.

Should we be populating user provided list-items on `new FastList()`, or are we happy to wait until render is next called?

File: fast-list.js (hunk 3 above)
Comment (etiennesegonzac) [inline fast-list.js:46]: :+1:

File: gaia-fast-list.js (hunk 6 above)
Comment (etiennesegonzac) [inline gaia-fast-list.js:187]: ha! like the naming :)

File: .jshintrc [hunk 12]
```diff
@@ -22,6 +22,7 @@
   "regexdash": true,
   "laxcomma": true,
   "proto": true,
+  "boss": true,
```
Comment (etiennesegonzac) [inline .jshintrc:25]: I never have any idea about what the jshint changed do, but I trust you :)

File: examples/gaia-fast-list/index.html [hunk 13]
```diff
@@ -15,18 +23,9 @@
...
         </li>
       </template>
     </gaia-fast-list>
-
-    <script src="bower_components/dom-scheduler/lib/dom-scheduler.js"></script>
-    <script src="bower_components/gaia-component/gaia-component.js"></script>
-    <script src="bower_components/poplar/poplar.js"></script>
-    <script src="fast-list.js"></script>
-    <script src="gaia-fast-list.js"></script>
-    <script src="app.js"></script>
```
Comment (etiennesegonzac) [inline examples/gaia-fast-list/index.html:30]: can you backport the manifest changes from #46 (adding the new component permission) to this PR so we get them landed fast?

File: examples/gaia-fast-list/app.js [hunk 14]
```diff
@@ -28,6 +28,26 @@ function loadNext() {
   });
 }
 
+function getDataSync(from, limit) {
```
Comment (etiennesegonzac) [inline examples/gaia-fast-list/app.js:29]: Is this to be kept around for debug? Can you add a small comment explaining it?

File: gaia-fast-list.js [hunk 15]
```diff
@@ -264,12 +329,20 @@ Internal.prototype = {
...
+    section.classList.add('gfl-section');
 
     return section;
   },
 
   populateItem: function(el, i) {
-    poplar.populate(el, this.getRecordAt(i));
+    // debug('populate item', el, i);
+    var record = this.getRecordAt(i);
+    if (!poplar.populate(el, record)) {
```
Comment (etiennesegonzac) [inline gaia-fast-list.js]: nit: not a huge fan of having an important action (populate) tucked away in a condition.
I'm guilty of this too, but I wouldn't mind the extra `poplarResult` var. (or a better name)

File: gaia-fast-list.js [hunk 16]
```diff
@@ -204,29 +252,39 @@ Internal.prototype = {
...
-    });
+      hash[section].push(items[i]);
+    }
 
     return hash;
   },
 
   createList: function() {
+    debug('create list');
     this.fastList = new FastList(this);
```
Comment (etiennesegonzac) [inline gaia-fast-list.js:285]: maybe we should just put those 2 lines in a `scheduler.mutation` block and remove the `setTimeout` in `updateCache`

File: examples/gaia-fast-list/app.js [hunk 17]
```diff
@@ -1,5 +1,8 @@
 'use strict';
 
+console.time('firstpaint');
```
Comment (etiennesegonzac) [inline examples/gaia-fast-list/app.js]: I don't see a `timeEnd` so we can probably remove this one :)

File: gaia-fast-list.js [hunk 18]
```diff
@@ -443,6 +521,30 @@ Internal.prototype = {
     });
   },
 
+  getCacheKey: function() {
+    return `${this.el.tagName}:${this.el.id}:${location}`;
+  },
+
+  updateCache: function() {
+    if (!this.el.caching) return;
```
Comment (etiennesegonzac) [inline gaia-fast-list.js:550]: `this.el.caching` or `this.caching`?

**Comment by etiennesegonzac** [comment]
Maybe the gaia-fast-list example should have `caching` on

File: gaia-fast-list.js (hunk 18 above)
Comment (wilsonpage) [inline gaia-fast-list.js:550]: `this.el` refers to the pubic `<gaia-fast-list>` element. `this.el.caching` refers to the public property/attribute on this node.

Comment (wilsonpage) [inline gaia-fast-list.js:550]: In this context `this` refers to the private `Internal` class.

**Comment by etiennesegonzac** [comment]
something goes wrong during one frame. not sure why (yet).
but it makes things jumpy.
here's a slow-mo gif :) 
vosuqz

**Comment by wilsonpage** [comment]
Oooo never seen that before :/

**Comment by wilsonpage** [comment]
Is this because you've refreshing the page a different scrollTop?

**Comment by etiennesegonzac** [comment]
> Is this because you've refreshing the page a different scrollTop?

I thought about this (we should support it anyway), but no that's not it.

File: gaia-fast-list.js (hunk 18 above)
Comment (etiennesegonzac) [inline gaia-fast-list.js:550]: got it, thanks

File: fast-list.js [hunk 19]
```diff
@@ -231,6 +234,7 @@ FastList.prototype = {
 
     function renderItem(i) {
       var item = items[i];
+
```
Comment (etiennesegonzac) [inline fast-list.js:237]: `source.populateItem` might replace the item from under us.
then `placeItem` gets passed the wrong dom node.

An easy fix is to do `item = items[i];` again just before calling `placeItem`.
Not very elegant, if you don't have a better idea I trust you to craft a very clear comment explaining it :)

Comment (wilsonpage) [inline fast-list.js:237]: Genius, good catch! Can you think or a more elegant way of replacing items? Are we happy to expect that users might use `.populateItem()` as a hook for switching out nodes?

File: gaia-fast-list.js (hunk 16 above)
Comment (wilsonpage) [inline gaia-fast-list.js:285]: But we want layerization to happen after the list has first been painted. Would putting it in `scheduler.mutation` block give Gecko enough time to do a quick paint before the heavy layerizing paint?

**Comment by wilsonpage** [comment]
> I thought about this (we should support it anyway), but no that's not it.

**Comment by wilsonpage** [comment]
> I thought about this (we should support it anyway), but no that's not it.

I'm not sure exactly what's required to support this. Would it involve caching the `scrollTop` and then calling `.reloadData()` when the current `scrollTop` doesn't match? Or perhaps we just always call reload data and hope that the `dataset.postion` checks prevent us unnecessarily rerendering anything.

**Comment by etiennesegonzac** [comment]
Yep, 2 nits in example/gaia-fast-list/app.js are still there, but otherwise we're good to go!
`populateItem` switching out node might be an issue but we can tackle that later.

File: gaia-fast-list.js (hunk 16 above)
Comment (etiennesegonzac) [inline gaia-fast-list.js:285]: good point, maybe just `updateCache` then (can be done inside the method, so it returns a scheduler promise)

**Comment by etiennesegonzac** [comment]
While we're updating updateCache (see scheduler comment): we might also want to limit the number of items in the cache.
We want a very fast first paint, so we should cache just the critical viewport.

File: gaia-fast-list.js (hunk 16 above)
Comment (wilsonpage) [inline gaia-fast-list.js:285]: Is that going to be deferred by enough time? A single microtask tick will mean that the `localStorage.setItem()` call will still block more important stuff.

**Comment by wilsonpage** [comment]
> While we're updating updateCache (see scheduler comment): we might also want to limit the number of items in the cache.
> We want a very fast first paint, so we should cache just the critical viewport.

Good point, is that number publicly available from `FastList`?

**Comment by wilsonpage** [comment]
We basically need a handle on the `criticalStart` and `criticalEnd` indices.

**Comment by etiennesegonzac** [comment]
Not really, but taking viewportheight / itemHeight (which we have) items + sections should do the trick

**Comment by etiennesegonzac** [comment]
The "not really" was in response to the first comment :)

**Comment by wilsonpage** [comment]
> Not really, but taking viewportheight / itemHeight (which we have) items + sections should do the trick

That approach works if we are sure the `scrollTop` is `0` at the time of caching. If there are list items above the viewport, there this will break.

**Comment by wilsonpage** [comment]
\o/

**Pull request merged**